Updated version of #2125

jonlives opened the pull request:
This PR takes @mdkent and @inno's work from #2125 and updates the specs to be compatible with the currently used version of RSpec so they should now pass. It also adds an extra unit test to hopefully handle @tyler-ball's comment in #2125 about calling .to_s on nil

This PR adds functionality the yum_package provider to support both specifying the version requirement in the package name, which is what was broken and reported in CHEF-2911, and also to support specifying a version attribute in the yum_package resource in one of the following two formats:

* Specific version, eg 3.0.1.el5
* Version requirement, eg = 2.0.1.el5

@chef/client-core if you'd like to cast your gaze over this revised version, I'll tackle any spec fails travis surfaces...they pass locally, but then that doesn't really mean anything :p

Documentation updates etc will be completed once this PR assumes its final form.

Eventually, it'd probably also be a good idea to deprecate supporting version requirements in the resource name and *just* use the version attribute - depending what people think, that could be incorporated in this PR.

btm commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1238,6 +1249,7 @@ def parse_dependency
             new_package_name = packages.first.name
+            new_package_version = packages.first.version.to_s
             Chef::Log.debug("#{@new_resource} no package found for #{@new_resource.package_name} " +
                             "but matched Provides for #{new_package_name}")
```
Might be nice for this to include a little more information, but it's just debugging data.

```
debug_msg = "#{@new_resource}: Unable to match package '#{@new_resource.package_name}' but matched #{packages.size} "
debug_msg << package.size == 1 ? "package" : "packages"
debug_msg << ", selected '#{new_package_name}' version '#{new_package_version}'"
Chef::Log.debug(debug_msg)

jonlives replied:
Good call, will add that

btm commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1220,7 +1220,18 @@ def parse_arch
+
+          # If we are passed a version we have to assume it's a requirement first. If it can't be
+          # parsed only yum_require.name will be set and @new_resource.version will be left intact
+          if @new_resource.version
```
add a comment that version is/may be a version constraint instead of just a version

jonlives replied:
done.

btm commented on spec/unit/provider/package/yum_spec.rb:
```diff
@@ -248,6 +248,93 @@
+      expect(@new_resource.version).to eq(nil)
+    end
+
+    it "should search provides if a package name can't be found then set version to match if a requirement was passed in the package name" do
```
It looks like there's a lot of duplication in these three tests. Start with moving @yum_cache to a let and put these in a context?

```
context "when the package name isn't found" do
  let(:yum_cache) { double(
          'Chef::Provider::Yum::YumCache',
          :reload_installed => true,
          :reset => true,
          :installed_version => "1.0.1.el5",
          :candidate_version => "2.0.1.el5",
          :package_available? => false,
          :version_available? => true,
          :disable_extra_repo_control => true
    )
  }

  before do
     allow(Chef::Provider::Package::Yum::YumCache).to receive(:instance).and_return(yum_cache)
     pkg = Chef::Provider::Package::Yum::RPMPackage.new("test-package", "2.0.1.el5", "x86_64", [])
     expect(yum_cache).to receive(:packages_from_require).and_return([pkg])
  end

  it "searches provides and sets version to match if a version constraint was passed in the version" do
    # ...
  end

  # ...
end
```

jonlives replied:
That makes sense I think - got a bunch of meetings this afternoon but will tackle that after or tomorrow morning if not.

jonlives commented:
@btm I've refactored the new tests inside a context, haven't touched any of the existing ones for now although there's still quite a lot of duplication lying around.

btm commented:
:+1: 

I've got a personal thing against the old cargo culted test pattern of ```it "should do this..."``` instead of ```it "does this"``` because the tests are an assertion, which I just mention to further my cause. :)

tyler-ball commented:
:+1:

lamont-granquist commented:
I think @jaymzh  has the write lock on the yum provider in #2692 -- we need to get that merged and should really rebase this on top of that.

jonlives commented:
@lamont-granquist I'm fine to wait for that if it's quick-ish, but this is fixing a bug (albeit a longstanding one) rather than adding a new feature so I'd like to get it in soon - looks like there some outstanding questions still on #2692 .

jonlives commented:
bias caveat: it's fixing a bug that we need fixed.

lamont-granquist commented:
I want the multipackage stuff in for 12.1.0 so all of this goes out in the same release anyway, and the rebase collisions this will cause will make phil's job a lot harder than it already is.

jonlives commented:
That's fair enough - easier for me to rebase this on top of Phil's work, and it doesn't hold up which release this goes out in.

lamont-granquist commented:
Okay, all clear for rebase + merge

jonlives commented:
Great, will tackle that first thing tomorrow.

jonlives commented:
@chef/client-core this is now rebased against @jaymzh's multi-package work.

@lamont-granquist / @jaymzh would particularly appreciate you two giving this the once-over, all specs are passing and I *think* I've preserved the original intent of your work but it'd be good to get some more eyes on this

jaymzh commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1275,7 +1277,18 @@ def parse_arch
+
+          # If we are passed a version or a version constraint we have to assume it's a requirement first. If it can't be
+          # parsed only yum_require.name will be set and @new_resource.version will be left intact
+          if @new_resource.version
```
these may be arrays in which case this won't work

jaymzh commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1275,7 +1277,18 @@ def parse_arch
+          # If we are passed a version or a version constraint we have to assume it's a requirement first. If it can't be
+          # parsed only yum_require.name will be set and @new_resource.version will be left intact
+          if @new_resource.version
+            require_string = "#{@new_resource.package_name} #{@new_resource.version}"
```
same

jaymzh commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1275,7 +1277,18 @@ def parse_arch
+          else
+            # Transform the package_name into a requirement, might contain a version, could just be
+            # a match for virtual provides
+            require_string = @new_resource.package_name
```
same

jonlives commented:
Thanks @jaymzh  - will take a look at fixing the above comments up and will try and add some additional specs to catch those cases

jaymzh commented:
FWIW, I did most of my testing with chef-apply, and things like:

```
package ['ack', 'less', 'iptstate'] do
  action :upgrade
end
```

and:

```
package ['ack', 'less', 'iptstate'] do
  version [...]
  action :install
end
```

And then uninstall some/all of those before I run the chef-apply, and/or install old versions of some.

I guess the goal here is to make something like this work:

```
package ['ack-<version>', 'less-<version>', 'iptstate-<version>'] do
  action :install
end
```
?

jonlives commented:
@jaymzh can you give an example of when the above might be an array and not work? parse_dependancy is called inside the iterator for package_name_array so surely we'd be dealing with one package at a time there?

jaymzh commented:
well, sure, when you use the `name` that was passed in. but when you reference `@new_resource.package_name`, you're looking back at the full resource and getting the full array. Same thing for `@new_resource.version`

jonlives commented:
@jaymzh aaah I get ya - rgr, will get that fixed and add some specs to test for it.

jonlives commented:
@jaymzh one more thing to double check. Is it a safe assumption that for every element in package_name_array there will be an equivalent entry in new_version_array, or it'll be blank? I'm going to need to pass the version number provided to parse_dependencies since I can't look at @new_resource and will of course need to make sure it's the correct version requirement.

jaymzh commented:
Yup. In fact we zip them together a lot. Missing versions will be nil.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jonlives commented:
@jaymzh / @chef/client-core I've updated to cater for @new_resource being an array and added a couple more specs to test this case.

lamont-granquist commented:
Yeah indexes on the *_array stuff all match up, and the superclass has a helper to step through them all in lock-step.

jaymzh commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1293,8 +1306,11 @@ def parse_dependency(name)
-            Chef::Log.debug("#{@new_resource} no package found for #{@new_resource.package_name} " +
-                            "but matched Provides for #{new_package_name}")
+            new_package_version = packages.first.version.to_s
+            debug_msg = "#{@new_resource}: Unable to match package '#{@new_resource.package_name}' but matched #{packages.size} "
```
that should be `...unable to match package '#{name}' but matched ...`

jaymzh commented on lib/chef/provider/package/yum.rb:
```diff
@@ -1054,15 +1054,17 @@ def load_current_resource
             unless @yum.package_available?(n)
               # If they aren't in the installed packages they could be a dependency
-              dep = parse_dependency(n)
+              dep = parse_dependency(n,new_version_array[index])
```
style nit - both ruby style and this code base, use `, ` between arguments to functions, but in the call and in the prototype, but in your diff, you consistently don't use the space (`foo(bar, baz)` vs `foo(bar,baz)`)

tyler-ball commented:
I would like to see a test showing that 
```ruby
package 'test-package = 2.0.1.el5' do
  version '2.0.1.el5' # or '3.0.0.el6'
end
```
fails sanely and with enough information for users to see what they did wrong.  It doesn't need to be a special message if the current error logging communicates this well enough.

tyler-ball commented:
Other than my minor comment and doc updates, :+1:

jonlives commented:
@tyler-ball currently, the parse method of the RPMDependency class would give precedence to the version requirement specified in the resource name - this is the way that worked before this patch too.

In terms of failure mechanism, I'm guessing you'd like to see an error if both are specified instead of the above behaviour?

tyler-ball commented:
@jonlives If it was not an error before, you don't need to change it to an error now.  But I think that test is still valuable for documentation purposes - it can just assert that the version specified in the resource name is the one that gets installed.

jonlives commented:
@tyler-ball I've added that final test per your request and updated the docs.

@lamont-granquist you're the other Enterprise Linux maintainer currently, would you be able to have a quick check over this please? Once everything's approved will rebase again off master before merging, there's a big ol' merge conflict resolution in this branch.

lamont-granquist commented:
One thing is you should probably squash your commits, the history will probably be a bit confusing at this point.

jonlives commented:
@lamont-granquist absolutely - will make sure it's all cleaned up for merge.

lamont-granquist commented:
Could you add a spec test to make sure that this works:

```ruby
yum_package ["foo", "bar"] do
  version ["1.0.1", "2.0.1"]
end
```

jonlives commented:
Sure, I can get that added tomorrow - prob should have been done as part of #2692.

jonlives commented:
@lamont-granquist added a spec testing that case, hopefully it's along the lines of what you were after.

tyler-ball commented:
:+1: :100: :cake: :gift:

lamont-granquist commented:
Sorta hoping to not only get load_current_resource, but see that the install action did the right stuff...

jonlives commented:
@lamont-granquist is that not covered by the existing tests that @jaymzh added for #2692 though? There are a bunch of tests there verifying what install_package does with arrays of names and versions.

jonlives commented:
I guess what I'm saying is I'm unclear where exactly we have a gap in test coverage here with the combination of @jaymzh's specs and the ones added in this PR. It's beginning to feel a bit like every day I wake up to another request for a spec and I'm never gonna get this merged so if possible it'd be good to bullet point the test cases I'm not catching here and I'll do my best to add specs for 'em - would love to get this in before the next version freeze :)

lamont-granquist commented:
Yeah, grok.  I'm in triage mtg right now.  I'll have to inspect the code closer when I get a bit of time.

jonlives commented:
Awesome thanks - much appreciated :)

lamont-granquist commented:
Okay yeah I get it, you only changed LCR codepaths, so if the output of that is correct, then its the responsibility of the rest of it to get the install right.

:+1:

jonlives commented:
Right - awesome, will get this rebased, squashed and merged by end of week. Thanks!

jonlives commented:
Ignore this, will open new PR with squashed version
